### PR TITLE
autopopulate user info

### PR DESCRIPTION
**Pull request by dandds**
Covers [this](https://www.pivotaltracker.com/story/show/158943860) PT story.

Any available user info relevant to the "Information About You" section of the requests form is mapped from the request creator to the request (first name, last name, email).

This required getting the user's email from the CAC certificate when available (when they use their DOD email cert versus their plainer DOD ID cert). The `/login-redirect` method got pretty gnarly, so I refactored the logic for authenticating a user into an `AuthenticationContext` object that encapsulates all the domain logic for our current auth flow. This involved some small changes to the CRL validator, too. The `AuthenticationContext` object now raises an `UnauthorizedError` wherever appropriate, and this will trigger our 401 redirect while raising context-specific errors.

### CODE REVIEW

File: tests/fixtures/artgarfunkel@uso.mil.crt [hunk 1]
```diff
@@ -0,0 +1,29 @@
+-----BEGIN CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (dandds) [reply tests/fixtures/artgarfunkel@uso.mil.crt]: I did. I used an `openssl` command. I'll add a short README to the fixtures directory about regenerating them.

**Comment by richard-dds** [comment]
@dandds Is there a way to prevent the log messages during testing? Now that we've been adding some logging I've seen a lot of noise in my pytest output.

**Comment by dandds** [comment]
@richard-dds that would take a PR in scriptz to remove the `-s` flag from pytest. I can do that later today.

**Comment by dandds** [comment]
@richard-dds Nevermind, got it.

Could one of you review again when you get a chance? (@patricksmithdds )

**Pull request merged**